app/src/main/kotlin/compiler/analysis/TypeChecker.kt

```diff
@@ -173,9 +173,33 @@ class TypeChecker(private val nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
                         is Function -> report(TypeCheckingError.FunctionAsValue(expression, node))
                     }
                 }
-                is Expression.ArrayElement -> TODO()
-                is Expression.ArrayLength -> TODO()
-                is Expression.ArrayAllocation -> TODO()
+
+                is Expression.ArrayElement -> {
+                    val innerExpressionType = checkExpression(expression.expression) ?: return null // if null then something must've gone bad in the recursion
+
+                    if (innerExpressionType is Type.Array) {
+                        checkExpression(expression.index, Type.Number)
```

Comment:
Don't we need failure handling here as well? :thinking: 
Add a test for this case too, something like `arr[prawda]`

Reply:
Yeee

Reply:
What failure handling specifically? `checkExpression` in 181 already appends to diagnostics if the type mismatches.
I'll add a test.

Reply:
You're right, we don't need to add anything here